modules/contact/src/constraints/TangentialLMConstraint.C

```diff
@@ -0,0 +1,229 @@
+  params.addCoupledVar("vel_z", "The z velocity");
+  params.addRequiredParam<Real>("mu", "The coefficient of friction.");
+  params.addParam<Real>("lambda", .95, "The weighting coefficient from Chen");
+
+  params.addClassDescription("Implements the KKT conditions for frictional Coulomb contact using "
+                             "an NCP function. Requires that either the relative tangential "
+                             "velocity is zero or the tangential stress is equal to the friction "
+                             "coefficient times the normal contact pressure.");
+  return params;
+}
+
+template <typename T>
+int
+sgn(T val)
+{
+  return (T(0) < val) - (val < T(0));
+}
+
+TangentialLMConstraint::TangentialLMConstraint(const InputParameters & parameters)
+  : NodeFaceConstraint(parameters),
+    _contact_pressure(getVar("contact_pressure", 0)->dofValues()),
+    _contact_pressure_id(coupled("contact_pressure")),
+
+    _vel_x(_master_var.dofValues()),
+    _vel_x_id(_master_var_num),
+
+    _vel_y(getVar("vel_y", 0)->dofValues()),
+    _vel_y_id(coupled("vel_y")),
+
+    _vel_z(isCoupled("vel_z") ? coupledValue("vel_z") : _zero),
+    _vel_z_id(isCoupled("vel_z") ? coupled("vel_z") : libMesh::invalid_uint),
+
+    _mu(getParam<Real>("mu")),
+    _lambda(getParam<Real>("lambda")),
+    _epsilon(std::numeric_limits<Real>::epsilon())
+{
+  _overwrite_slave_residual = false;
+}
+
+Real
+TangentialLMConstraint::computeQpSlaveValue()
+{
+  return _u_slave[_qp];
+}
+
+void
+TangentialLMConstraint::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+
+  _qp = 0;
+
+  re(0) = computeQpResidual(Moose::Slave);
+}
+
+void
+TangentialLMConstraint::computeJacobian()
+{
+  _Kee.resize(1, 1);
+  _connected_dof_indices.clear();
+  _connected_dof_indices.push_back(_var.nodalDofIndex());
+
+  _qp = 0;
+
+  _Kee(0, 0) += computeQpJacobian(Moose::SlaveSlave);
+}
+
+void
+TangentialLMConstraint::computeOffDiagJacobian(unsigned jvar)
+{
+  if (jvar == _var.number())
+  {
+    computeJacobian();
+    return;
+  }
+
+  MooseVariableFEBase & var = _sys.getVariable(0, jvar);
+  _connected_dof_indices.clear();
+  _connected_dof_indices.push_back(var.nodalDofIndex());
+
+  _qp = 0;
+
+  _Kee.resize(1, 1);
+  _Kee(0, 0) += computeQpOffDiagJacobian(Moose::SlaveSlave, jvar);
+}
+
+Real TangentialLMConstraint::computeQpResidual(Moose::ConstraintType /*type*/)
+{
+  std::map<dof_id_type, PenetrationInfo *>::iterator found =
+      _penetration_locator._penetration_info.find(_current_node->id());
+  if (found != _penetration_locator._penetration_info.end())
+  {
+    PenetrationInfo * pinfo = found->second;
+    if (pinfo != NULL)
+    {
+      if (_contact_pressure[_qp] < _epsilon)
+        return _u_slave[_qp];
+      else
+      {
+        RealVectorValue tangent_vec(pinfo->_normal(1), -pinfo->_normal(0), 0);
+        Real v_dot_tan = RealVectorValue(_vel_x[_qp], _vel_y[_qp], _vel_z[_qp]) * tangent_vec;
```

Comment:
Can't we just use u - u_old here instead of requiring the user to couple in a velocity variable?